presto-main/src/main/java/com/facebook/presto/sql/planner/SortChannelExtractor.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FieldReference;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+public class SortChannelExtractor
+{
+    public static class SortChannel
+    {
+        private final int channel;
+        private final boolean descending;
+
+        public SortChannel(int channel, boolean descending)
+        {
+            this.channel = channel;
+            this.descending = descending;
+        }
+
+        public int getChannel()
+        {
+            return channel;
+        }
+
+        public boolean isDescending()
+        {
+            return descending;
+        }
+    }
+
+    private SortChannelExtractor()
+    {
+    }
+
+    public static Optional<SortChannel> getSortChannel(Map<Symbol, Integer> buildLayout, Expression filter)
+    {
+        Set<Integer> buildFields = ImmutableSet.copyOf(buildLayout.values());
+        if (filter instanceof ComparisonExpression) {
+            ComparisonExpression comparison = (ComparisonExpression) filter;
+            boolean descending = false;
+            switch (comparison.getType()) {
+                case EQUAL:
+                case NOT_EQUAL:
+                case IS_DISTINCT_FROM:
+                    break;
+                case GREATER_THAN:
+                case GREATER_THAN_OR_EQUAL:
+                    descending = true;
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                    Optional<Integer> sortChannel = isBuildFieldReference(buildFields, comparison.getRight());
+                    if (!sortChannel.isPresent()) {
+                        sortChannel = isBuildFieldReference(buildFields, comparison.getLeft());
+                        descending = !descending;
+                    }
+                    if (sortChannel.isPresent()) {
+                        return Optional.of(new SortChannel(sortChannel.get(), descending));
+                    }
+                    return Optional.empty();
```

Comment:
yes, I have missed one thing. For such simple extractor, I have to check that there is no build field references on the other side of the inequality. Fixed